### PR TITLE
Update azure-pipelines.ts

### PR DESCRIPTION
**Pull request by deepdotnet**
Added azure build, release, and deployment versions. Support for private azure projects.

### CODE REVIEW

**Comment by deepdotnet** [comment]
I think it's not breaking change. I have kept old URL pattern as well for backword compatibility.

**Comment by tunnckoCore** [comment]
ah yea probably, im a bit sleepy

File: api/azure-pipelines.ts [hunk 1]
```diff
@@ -1,42 +1,130 @@
...
   title: 'Azure Piplines',
   examples: {
-    '/azure-pipelines/yarnpkg/yarn/Yarn Acceptance Tests': 'build',
-    '/azure-pipelines/yarnpkg/yarn/Yarn Acceptance Tests/azure-pipelines': 'build (branch)',
+    '/azure-pipelines/build/status/yarnpkg/yarn/Yarn Acceptance Tests': 'build',
+    '/azure-pipelines/build/status/yarnpkg/yarn/Yarn Acceptance Tests/azure-pipelines': 'build (branch)',
   },
   handlers: {
-    '/azure-pipelines/:org/:project/:definition/:branch?': handler
+    '/azure-pipelines/:org/:project/:definition/:branch?': buildStatus,
```
Comment (amio) [inline api/azure-pipelines.ts]: This better be placed last as a deprecated fallback

File: api/azure-pipelines.ts [hunk 2]
```diff
@@ -1,42 +1,130 @@
...
-  const response = await got(`https://dev.azure.com/${org}/${project}/_apis/build/status/${definition}?branchName=${branch}`, { json: false })
-  const contentType = response.headers['content-type']
-
-  if (!contentType.includes('image/svg+xml')) {
-    return {
-      subject: 'Azure Pipelines',
-      status: 'unknown',
-      color: 'grey'
-    }
+const AZURE_DEVOPS_TOKEN = process.env.AZURE_DEVOPS_TOKEN
```
Comment (amio) [inline api/azure-pipelines.ts]: Need to add this to [`now.json`](https://zeit.co/docs/configuration/#project/env):

```diff
	{
	  "env": {
	    "GH_TOKENS": "@badgen-gh-tokens",
	    "SENTRY_DSN": "@badgen-sentry-dsn",
	    "TRACKING_GA": "@badgen-tracking-ga",
+	    "AZURE_DEVOPS_TOKEN": "@badgen-azure-token"
	  }
	}
```

File: api/azure-pipelines.ts [hunk 3]
```diff
@@ -1,42 +1,130 @@
...
 import { createBadgenHandler, PathArgs } from '../libs/create-badgen-handler'
+import { isNullOrUndefined } from 'util'
 
 export default createBadgenHandler({
   title: 'Azure Piplines',
   examples: {
-    '/azure-pipelines/yarnpkg/yarn/Yarn Acceptance Tests': 'build',
-    '/azure-pipelines/yarnpkg/yarn/Yarn Acceptance Tests/azure-pipelines': 'build (branch)',
+    '/azure-pipelines/build/status/yarnpkg/yarn/Yarn Acceptance Tests': 'build',
+    '/azure-pipelines/build/status/yarnpkg/yarn/Yarn Acceptance Tests/azure-pipelines': 'build (branch)',
```
Comment (amio) [inline api/azure-pipelines.ts]: It would be nice to have more examples showcasing all these features 👍

File: api/azure-pipelines.ts [hunk 4]
```diff
@@ -1,42 +1,130 @@
...
+  const testResult = await getTestResultByBuild({org, project, build: build.id, minLastUpdatedDate: build.startTime, maxLastUpdatedDate: build.finishTime})
+  const runStatistics = testResult.value[0].runStatistics
+  const total: number = testResult.value[0].totalTests
+  const passed: {outcome: string, count: number} = runStatistics.find( (value: { outcome: string; }) => value.outcome === 'Passed')
+  const notExecuted: {outcome: string, count: number} = runStatistics.find( (value: { outcome: string; }) => value.outcome === 'NotExecuted')
+  const failed: {outcome: string, count: number} = runStatistics.find( (value: { outcome: string; }) => value.outcome === 'Failed')
+  
+  const passedCount = !isNullOrUndefined(passed) ? passed.count : 0
+  const notExecutedCount = !isNullOrUndefined(notExecuted) ? notExecuted.count : 0
+  const failedCount =  !isNullOrUndefined(failed) ? failed.count : total - passedCount - notExecutedCount
```
Comment (amio) [inline api/azure-pipelines.ts]: We are using typescript ^3.7 now, these can be simplified to:

```js
  const passedCount = passed?.count ?? 0
  const notExecutedCount = notExecuted?.count ?? 0
  const failedCount = failed?.count ?? total - passedCount - notExecutedCount
```

Comment (deepdotnet) [reply api/azure-pipelines.ts]: Thanks for the suggestion. Implemented the same.

File: api/azure-pipelines.ts (hunk 3 above)
Comment (deepdotnet) [reply api/azure-pipelines.ts]: Added build version example. I don't have any Azure DevOps public project which I can use as an example. Need to figure out how to put examples.

File: api/azure-pipelines.ts (hunk 1 above)
Comment (deepdotnet) [reply api/azure-pipelines.ts]: Done

**Comment by deepdotnet** [comment]
I have added the old handler back as new code doesn't work with the definition name but with ID only.

**Comment by deepdotnet** [comment]
@amio what needs to be done for above failure?

**Comment by amio** [comment]
I need to add a azure devops token to now secrets.

@deepdotnet Is this token required? I tried locally, it works without a token.

File: api/azure-pipelines.ts (hunk 3 above)
Comment (amio) [reply api/azure-pipelines.ts]: @deepdotnet Can we use `yarnpkg/yarn` for all examples?

File: api/azure-pipelines.ts [hunk 5]
```diff
@@ -1,18 +1,136 @@
-import cheerio from 'cheerio'
```
Comment (amio) [inline api/azure-pipelines.ts]: `cheerio` need to be imported

Comment (deepdotnet) [reply api/azure-pipelines.ts]: The token is not mandatory. The token is required for private projects.

**Comment by tunnckoCore** [comment]
> The token is not mandatory. The token is required for private projects.

He could add it, but it will be for _his_ private repos i think? I'm not familiar much with azure.

File: api/azure-pipelines.ts (hunk 3 above)
Comment (deepdotnet) [reply api/azure-pipelines.ts]: Yes, I tried, I didn't find any release info in API response. If we can add example UTs, ATs, and release in yarnpkg/yarn, I can add examples here.

**Comment by deepdotnet** [comment]
> 
> 
> > The token is not mandatory. The token is required for private projects.
> 
> He could add it, but it will be for _his_ private repos i think? I'm not familiar much with azure.

Yes, Azure DevOps APIs require an access token for private projects, however public projects can be accessed without an access token. Certain APIs require an access token even for public projects. Having the capability to provide access token in badgen is useful for people looking for on-premise/their own deployment.

**Comment by amio** [comment]
> Azure DevOps APIs require an access token for private projects

@deepdotnet I thought it's required for requesting azure api. If we want to support private project, the token should be provided by user, but it's not a good idea to let user put their token in badge url, so we don't support private project on badgen.net.

We need to remove the env config from `now.json` 😅

File: api/azure-pipelines.ts (hunk 3 above)
Comment (amio) [reply api/azure-pipelines.ts]: At least we would need a working example to test if the handler works as expect.

**Comment by tunnckoCore** [comment]
> If we want to support private project, the token should be provided by user

Yea, exactly what I meant.

**Comment by deepdotnet** [comment]
> 
> 
> > Azure DevOps APIs require an access token for private projects
> 
> @deepdotnet I thought it's required for requesting azure api. If we want to support private project, the token should be provided by user, but it's not a good idea to let user put their token in badge url, so we don't support private project on badgen.net.
> 
> We need to remove the env config from `now.json` 😅

For the same reason, I initially didn't add it :). Will remove it again.

**Comment by deepdotnet** [comment]
Removed Azure token from now.

**Comment by amio** [comment]
> For the same reason, I initially didn't add it :). Will remove it again.

My bad 😝

**Pull request merged**